File: whatsapp_gateway/src/utils/phone.ts
/**
 * E.164 phone number normalization utility.
 * Extracts phone numbers from Baileys JIDs and normalizes to E.164 format.
 */

const INDIVIDUAL_SUFFIX = "@s.whatsapp.net";
const GROUP_SUFFIX = "@g.us";

/**
 * Normalizes a Baileys JID to E.164 format.
 *
 * - For individual chats (`<number>@s.whatsapp.net`), strips the suffix and prepends `+`.
 * - Validates the result is 8–15 digits (per E.164 spec).
 * - Returns `null` for group JIDs (`@g.us`) or invalid numbers.
 * - Idempotent: calling on an already-normalized E.164 string returns the same value.
 */
export function normalizeToE164(jid: string): string | null {
  if (!jid || typeof jid !== "string") {
    return null;
  }

  // Reject group JIDs
  if (jid.endsWith(GROUP_SUFFIX)) {
    return null;
  }

  let numberPart: string;

  if (jid.endsWith(INDIVIDUAL_SUFFIX)) {
    // Extract number from JID format: <number>@s.whatsapp.net
    numberPart = jid.slice(0, -INDIVIDUAL_SUFFIX.length);
  } else if (jid.startsWith("+")) {
    // Already in E.164 format (idempotent case)
    numberPart = jid.slice(1);
  } else {
    // Bare number without suffix or prefix
    numberPart = jid;
  }

  // Strip any non-digit characters
  const digits = numberPart.replace(/\D/g, "");

  // Validate E.164: 8–15 digits
  if (digits.length < 8 || digits.length > 15) {
    return null;
  }

  return `+${digits}`;
}

File: app/utils/phone.py
"""Phone number normalization utility.

Normalizes raw phone strings to E.164 format using the phonenumbers library.
Returns None for invalid inputs. Idempotent for valid E.164 numbers.
"""

from __future__ import annotations

import phonenumbers


def normalize_to_e164(raw: str) -> str | None:
    """Normalize a raw phone string to E.164 format.

    Args:
        raw: A raw phone number string (e.g., "081234567890", "+6281234567890").

    Returns:
        The E.164 formatted string (e.g., "+6281234567890") if valid,
        or None if the input cannot be parsed or is not a valid phone number.

    The function is idempotent: calling it on an already-normalized E.164 string
    returns the same string unchanged.
    """
    if not isinstance(raw, str) or not raw.strip():
        return None

    raw = raw.strip()

    try:
        # If the number starts with '+', parse without a default region
        # Otherwise, assume Indonesia (ID) as default region
        if raw.startswith("+"):
            parsed = phonenumbers.parse(raw, None)
        else:
            parsed = phonenumbers.parse(raw, "ID")
    except phonenumbers.NumberParseException:
        return None

    if not phonenumbers.is_valid_number(parsed):
        return None

    formatted = phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)

    # Validate E.164 length constraint: 8 to 15 digits total (per Req 2.1)
    digits = formatted[1:]  # strip leading '+'
    if not (8 <= len(digits) <= 15):
        return None

    return formatted
